#!/bin/bash
#---------------------------------------------------------------------------------------------------------
# run_variant_hardfilter.sh
#---------------------------------------------------------------------------------------------------------

# tool for filtering vcf files generated by GATK Unified Genotyper. Presumably it will work for
# VCF files generated by other SNP calling tools, however, they are not tested. This is a very hard filter
# and could be relaxed, particularly in the VCFtools section.
# GATK filters: QD < 2.0 || FS > 60.0 || MQ < 40.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0
# VCFtools filters: no missing data, biallelic SNPs

# Requirements: GATK, vcftools
# Input files: reference sequence, VCF file

# Use: ./run_variant_hardfilter.sh <out_prefix> <ref_fasta_full-path> <vcf_file>

# sd21@sanger.ac.uk
# V1: April 2016
# V2: Feb 2023

# Filters applied:
# --filterExpression ( filters applied to INFO field )
# QD: Variant Confidence/Quality by Depth = < 2
# FS: Phred-scaled p-value using Fisher's exact test to detect strand bias = > 60
# MQ: RMS Mapping Quality = < 40
# MQRankSum: Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities = < 40
# ReadPosRankSum: = < -8
#
# --genotypeFilterExpression ( filters applied in FORMAT field )
# DP: Approximate read depth (reads with MQ=255 or with bad mates are filtered) = < 10
# GQ: Genotype Quality = < 30
#
# -- vcftools
# - remove indels ( --remove-indels )
# - biallelic SNPs only (--min-alleles 2 --max-alleles 2 )
# - remove all SNPs and genotypes with filter flags set by GATK ( --remove-filtered-all --remove-filtered-geno-all )
# - no missing data, ie. all individuals for a given SNP must have a genotype ( --max-missing 1 )


##############################################################################################
module load gatk/4.1.4.1
module load vcftools/0.1.16-c4


PREFIX=$1
REFERENCE=$2
VCF=$3

if [ "$1" == "-h" ]; then
  echo "Usage: ./run_variant_hardfilter.sh <out_prefix> <ref_fasta_full-path> <vcf_file>"
  exit 0
  else
gatk VariantFiltration \
-R ${REFERENCE} \
--variant ${VCF} \
--filter-name "FILTERED" \
--filter-expression "QD < 2.0 || FS > 60.0 || MQ < 40.0 || ReadPosRankSum < -8.0" \
--genotype-filter-name "GTFILTER" \
--genotype-filter-expression "DP < 10" \
--output ${PREFIX}.filtered-1.vcf

vcftools-0.1.14 \
--vcf ${PREFIX}.filtered-1.vcf \
--max-missing 1 \
--remove-filtered-geno-all \
--remove-filtered-all \
--remove-indels \
--min-alleles 2 \
--max-alleles 2 \
--recode \
--out ${PREFIX}.filtered-2.vcf
fi