
# Guinea worm genetic map



# working directory: /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP


# RAW data: /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/RAW


# Samples - samples_lanes.list

# DM_L1_DOG_001	26584_1_1
# DM_L1_DOG_002	26584_1_2
# DM_L1_DOG_003	26584_1_3
# DM_L1_DOG_004	26584_1_4
# DM_L1_DOG_005	26584_1_5
# DM_L1_DOG_006	26584_1_6
# DM_L1_DOG_007	26584_1_7
# DM_L1_DOG_008	26584_1_8
# DM_L1_DOG_009	26584_1_9
# DM_L1_DOG_010	26584_1_10
# DM_L1_DOG_011	26584_1_11
# DM_L1_DOG_012	26584_1_12
# DM_L1_DOG_013	26584_1_13
# DM_L1_DOG_014	26584_1_14
# DM_L1_DOG_015	26584_1_15
# DM_L1_DOG_016	26584_1_16
# DM_L1_DOG_017	26584_1_17
# DM_L1_DOG_018	26584_1_18
# DM_L1_DOG_019	26584_1_19
# DM_L1_DOG_020	26584_1_20
# DM_L1_DOG_021	26584_1_21
# DM_L1_DOG_022	26584_1_22
# DM_L1_DOG_023	26584_1_23
# DM_L1_DOG_024	26584_1_24
# DM_L1_DOG_025	26584_1_25
# DM_L1_DOG_026	26584_1_26
# DM_L1_DOG_027	26584_1_27
# DM_L1_DOG_028	26584_1_28
# DM_L1_DOG_029	26584_1_29
# DM_L1_DOG_030	26584_1_30
# DM_L1_DOG_031	26584_1_31
# DM_L1_DOG_032	26584_1_32
# DM_L1_DOG_033	26584_1_33
# DM_L1_DOG_034	26584_1_34
# DM_L1_DOG_035	26584_1_35
# DM_L1_DOG_036	26584_1_36
# DM_L1_DOG_037	26584_1_37
# DM_L1_DOG_038	26584_1_38
# DM_L1_DOG_039	26584_1_39
# DM_L1_DOG_040	26584_1_40
# DM_L1_DOG_041	26584_1_41
# DM_L1_DOG_042	26584_1_42
# DM_L1_DOG_043	26584_1_43
# DM_L1_DOG_044	26584_1_44
# DM_L1_DOG_045	26584_1_45
# DM_L1_DOG_046	26584_1_46
# DM_L1_DOG_047	26584_1_47
# DM_L1_DOG_048	26584_1_48
# DM_L1_DOG_049	26584_1_49
# DM_L1_DOG_050	26584_1_50
# DM_L1_DOG_051	26584_1_51
# DM_L1_DOG_052	26584_1_52
# DM_L1_DOG_053	26584_1_53
# DM_L1_DOG_054	26584_1_54
# DM_L1_DOG_055	26584_1_55
# DM_L1_DOG_056	26584_1_56
# DM_L1_DOG_057	26584_1_57
# DM_L1_DOG_058	26584_1_58
# DM_L1_DOG_059	26584_1_59
# DM_L1_DOG_060	26584_1_60
# DM_L1_DOG_061	26584_1_61
# DM_L1_DOG_062	26584_1_62
# DM_L1_DOG_063	26584_1_63
# DM_L1_DOG_064	26584_1_64
# DM_L1_DOG_065	26584_1_65
# DM_L1_DOG_066	26584_1_66
# DM_L1_DOG_067	26584_1_67
# DM_L1_DOG_068	26584_1_68
# DM_L1_DOG_069	26584_1_69
# DM_L1_DOG_070	26584_1_70
# DM_L1_DOG_071	26584_1_71
# DM_L1_DOG_072	26584_1_72
# DM_L1_DOG_073	26584_1_73
# DM_L1_DOG_074	26584_1_74
# DM_L1_DOG_075	26289_8_1
# DM_L1_DOG_076	26289_8_2
# DM_L1_DOG_077	26289_8_3
# DM_L1_DOG_078	26289_8_4
# DM_L1_DOG_079	26289_8_5
# DM_L1_DOG_080	26289_8_6
# DM_L1_DOG_081	26289_8_7
# DM_L1_DOG_082	26289_8_8
# DM_L1_DOG_083	26300_1_1
# DM_L1_DOG_084	26289_8_9
# DM_L1_DOG_085	26300_1_2
# DM_L1_DOG_086	26289_8_10
# DM_L1_DOG_087	26289_8_11
# DM_L1_DOG_088	26289_8_12
# DM_L1_DOG_089	26289_8_13
# DM_L1_DOG_090	26289_8_14
# DM_L1_DOG_091	26300_1_3
# DM_L1_DOG_092	26300_1_4
# DM_L1_DOG_093	26289_8_15
# DM_L1_DOG_094	26289_8_16
# DM_L1_DOG_095	26289_8_17
# DM_L1_DOG_096	26289_8_18
# DM_L1_DOG_097	26300_1_5
# DM_L1_DOG_098	26300_1_6
# DM_L1_DOG_099	26300_1_7
# DM_L1_DOG_100	26300_1_8
# DM_L1_DOG_101	26289_8_19
# DM_L1_DOG_102	26300_1_9
# DM_L1_DOG_103	26289_8_20
# DM_L1_DOG_104	26300_1_10
# DM_L1_DOG_105	26289_8_21
# DM_L1_DOG_106	26289_8_22
# DM_L1_DOG_107	26300_1_11
# DM_L1_DOG_108	26300_1_12
# DM_L1_DOG_109	26289_8_23
# DM_L1_DOG_110	26289_8_24
# DM_L1_DOG_111	26289_8_25
# DM_L1_DOG_112	26289_8_26
# DM_L1_DOG_113	26289_8_27
# DM_L1_DOG_114	26300_1_13
# DM_L1_DOG_115	26289_8_28
# DM_L1_DOG_116	26289_8_29
# DM_L1_DOG_117	26289_8_30
# DM_L1_DOG_118	26289_8_31
# DM_L1_DOG_119	26300_1_14
# DM_L1_DOG_120	26289_8_32
# DM_L1_DOG_121	26289_8_33
# DM_L1_DOG_NEG_01	26289_8_34
# DM_L1_DOG_122	26289_8_35
# DM_L1_DOG_123	26289_8_36
# DM_L1_DOG_124	26289_8_37
# DM_L1_DOG_POS_2XL1	26300_1_15
# DM_L1_DOG_125	26289_8_38
# DM_L1_DOG_126	26289_8_39
# DM_L1_DOG_127	26289_8_40
# DM_L1_DOG_128	26300_1_16
# DM_L1_DOG_129	26289_8_41
# DM_L1_DOG_130	26289_8_42
# DM_L1_DOG_131	26289_8_43
# DM_L1_DOG_132	26289_8_44
# DM_L1_DOG_133	26289_8_45
# DM_L1_DOG_134	26289_8_46
# DM_L1_DOG_135	26289_8_47
# DM_L1_DOG_NEG_02	26289_8_48
# DM_L1_DOG_136	26289_8_49
# DM_L1_DOG_137	26289_8_50
# DM_L1_DOG_138	26289_8_51
# DM_L1_DOG_139	26289_8_52
# DM_L1_DOG_140	26289_8_53
# DM_L1_DOG_141	26289_8_54
# DM_L1_DOG_142	26289_8_55
# DM_L1_DOG_143	26289_8_56
# DM_L1_DOG_144	26289_8_57
# DM_L1_DOG_145	26289_8_58



mkdir 00_SCRIPTS
mkdir 01_REFERENCE
mkdir 02_RAW
mkdir 03_MAPPING



################################################################################################
# Step 1 -reference
################################################################################################
cd 01_REFERENCE

cp /lustre/scratch118/infgen/team133/jc17/Guinea_worm/New_SNP_calls/Dracunculus_medinensis_v2.QC.fa .
samtools faidx Dracunculus_medinensis_v2.QC.fa
samtools dict Dracunculus_medinensis_v2.QC.fa > Dracunculus_medinensis_v2.QC.dict

################################################################################################
# Step 2 - get and process raw data
################################################################################################
# process raw reads
cd /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/02_RAW

# --- check for fastqc and multiqc

multiqc_report.html


# trim reads
mkdir TRIM
cd TRIM

while read name lane; do \
echo "bsub.py --threads 4 10 trim ~sd21/bash_scripts/run_trimmomatic ${name} ../${lane}_1.fastq.gz ../${lane}_2.fastq.gz" >> run_trim_all; \
done < ../samples_lanes.list ; \
chmod a+x run_trim_all; \
./run_trim_all

# get rid of the unpaired fraction
rm *unpaired*

# make a sample list for mapping
ls -1 *.paired_R1.fastq.gz | sed 's/.paired_R1.fastq.gz//g' > ../../03_MAPPING/sample.list

################################################################################################
# Step 3 - Mapping
################################################################################################
cd /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/03_MAPPING

screen

while read NAME; do ~sd21/bash_scripts/run_bwamem_splitter ${NAME} /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/01_REFERENCE/Dracunculus_medinensis_v2.QC.fa /
nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/02_RAW/TRIM/${NAME}.paired_R1.fastq.gz /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/02_RAW/TRIM/${NAME}.pai
red_R2.fastq.gz; done < sample.list

# once completed
multiqc .

ls -1 $PWD/*out/*.merged.sorted.marked.bam > ../04_SNPS/bams.list


################################################################################################
# Step 4 - SNP calling and intial filtering
################################################################################################

./run_GATK_UG_fast DM_GeneticMap /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/01_REFERENCE/Dracunculus_medinensis_v2.QC.fa bam.list

# merge vcfs
grep ">" /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/01_REFERENCE/Dracunculus_medinensis_v2.QC.fa | sed -e 's/>//g' -e 's/$/.raw.tmp.vcf/g' > vcf.list
vcf-concat -f vcf.list > DM_GeneticMap.raw.vcf

# raw stats - vcftools
# After filtering, kept 149 out of 149 Individuals
# After filtering, kept 664977 out of a possible 664977 Sites
# Run Time = 9.00 seconds


# clean up
rm *tmp* gatk*

# filter
#--- step1
java -jar $gatk_dir/GenomeAnalysisTK.jar -T VariantFiltration -R /nfs/users/nfs_s/sd21/lustre118_link/dm/GENETIC_MAP/01_REFERENCE/Dracunculus_medinensis_v2.QC.fa --v
ariant DM_GeneticMap.raw.vcf --filterName "FILTERED" --filterExpression "QD < 2.0 || FS > 60.0 || MQ < 40.0 || ReadPosRankSum < -8.0 || StrandOddsRatio > 3" -o DM_Ge
neticMap.filtered-1.vcf
#--- step2
vcftools-0.1.14 --vcf DM_GeneticMap.filtered-1.vcf --max-missing 0.75 --remove-filtered-geno-all --remove-filtered-all --remove-indels --min-alleles 2 --max-alleles
2 --recode --recode-INFO-all --out DM_GeneticMap.filtered-2.vcf

# filtered stats
# After filtering, kept 149 out of 149 Individuals
# After filtering, kept 399741 out of a possible 664996 Sites
# Run Time = 36.00 seconds


################################################################################################
# Step 5 - SNP filtering for genetic map
################################################################################################
# AIM : heterozygous sites in adult female, and segregating sites in progeny

# determine the genotype frequency and concordance for the parental SNPs from a HWE analysis from vcftools
vcftools-0.1.14 --vcf DM_GeneticMap.filtered-2.vcf.recode.vcf --indv DM_AdultF_DOG_001 --hardy --out parental

# Summary data
# awk '{if($3=="1/0/0")  print $1,$2}' parental.hwe | wc -l
#>> 136364
# awk '{if($3=="0/0/1")  print $1,$2}' parental.hwe | wc -l
#>> 61911
# awk '{if($3=="0/1/0")  print $1,$2}' parental.hwe | wc -l
#>> 201400

# Extract SNPs that are heterozygous in the parental sampels
awk '{if($3=="0/1/0")  print $1,$2}' parental.hwe > parental_het.snplist

# extract SNPs that are homozygous in the parental sampels
awk '{if($3=="1/0/0" || $3=="0/0/1") print $1,$2}' parental.hwe > parental_homo.snplist

# Generate new SNP vcf that contains only SNPs in which the female is heterozygous  at that position
vcftools-0.1.14 --vcf DM_GeneticMap.filtered-2.vcf.recode.vcf --keep L1_progeny.list --positions parental_het.snplist --recode --recode-INFO-all --out DM_L1_femhets

# Generate new SNP vcf that contains only SNPs in which the female is homozygous at that position
vcftools-0.1.14 --vcf DM_GeneticMap.filtered-2.vcf.recode.vcf --keep L1_progeny.list --positions parental_homo.snplist --recode --recode-INFO-all --out DM_L1_femhomo


# determine SNP types from female het SNPs
./run_F1_genotypes_2_crosstype DM_L1_femhet DM_L1_femhets.recode.vcf L1_progeny.list

# determine SNP types from female homo SNPs
./run_F1_genotypes_2_crosstype DM_L1_femhomo DM_L1_femhomo.recode.vcf L1_progeny.list


# Summary data
# >> 3740 DM_L1_femhet.backcross_SNPs.list
# >> 1964 DM_L1_femhet.intercross_SNPs.list
# >> 195696 DM_L1_femhet.nonsegregating_SNPs.list
# >> 986 DM_L1_femhomo.backcross_SNPs.list
# >> 80 DM_L1_femhomo.intercross_SNPs.list
# >> 197209 DM_L1_femhomo.nonsegregating_SNPs.list


# Extract and generate new vcf of backcross SNPs where female is heterozygous
vcftools-0.1.14 --vcf DM_L1_femhets.recode.vcf --positions DM_L1_femhet.backcross_SNPs.list --recode --recode-INFO-all --out DM_L1_femhets.backcross
#>> F1X_L3_femhets.backcross.recode.vcf


# Generate RQTL files for  female het  sites that have a backcross segregation pattern, and thinned to 1 SNP per 5000-bp
./run_vcf_2_rqtl_bc DM_L1_femhets.backcross DM_L1_femhets.backcross.recode.vcf 5000

# After filtering, kept 1566 out of a possible 3740 Sites
# Run Time = 0.00 seconds
